### ISSUE TITLE
unbanlanced sparsity across the transformer layers

### ISSUE THREAD

**Issue by xuych0826** [thu-ml/ReMoE#6]
Congratulations! Your creative work has been accepted by ICRL 2025.

However, while applying this method to my own project, I observed that in certain transformer layers, the sparsity reaches 100% (despite the overall sparsity is closed to the target sparsity). This means that the routing map is entirely False, and all probability values (_probs_ in the code) are 0.0. This issue frequently occurs during the initial stages of my model training and leads to errors. 

Therefore, I would like to inquire whether this situation is normal, or this is just a issue of the initial lambda value and the multiplier value? 

If the sparsity of a particular layer is 100%, does it imply that we should bypass the feedforward layer entirely? 

Below are some training data I collected:

![Image](https://github.com/user-attachments/assets/d333d785-b901-41c7-8b51-1cdfcb722620)

![Image](https://github.com/user-attachments/assets/61576658-2fd3-4401-aa19-96c070517829)

![Image](https://github.com/user-attachments/assets/d79806b6-7425-458f-9bda-0a125089f654)

**Comment by thuwzt** [comment]
Hi, 

Thanks for for your interest in our work!

In our experiments, we haven’t encountered the "entirely zero layer" issue, even under extremely sparse configurations (1 in 128, ~99.2% sparsity).

This might be related to irregular data batches (which could explain the spikes in your sparsity plot—something we haven’t observed either), or simply a small micro batch size.

To help investigate further, could you share additional details about the model modality,  model structure, loss function, sequence length and micro batch size?

**Comment by xuych0826** [comment]
In my project, I am working with a diffusion transformer, where the transformer is utilized to predict the noise added to the original tokens.

The loss function employed is the Mean Squared Error (MSE) loss between the predicted noise and the actual noise.

Given that it is a relatively small model, I have not incorporated any techniques related to distributed training. The configuration includes a batch size of 256, a sequence length of 10, and a hidden layer size of 256.  Additionally, I am not utilizing micro-batching; hence, you can consider the number of micro-batches to be 1.

I don't think the batches is irregular and the size remains (256, 10, 256) all the time

The number of experts is 4, and topK = 1. 

Additionally, I found that, in the third stage, each time the overall sparsity is about to reach the target sparisity(75%), the issue occurs.

**Comment by xuych0826** [comment]
Oh, sorry!
I might found out why!
The shape of the last batch for each epoch is (20, 10, 256). Which means this issue occurs because the batch size is too small, thanks for your helpful reply!
